[Fiber] Push class context providers even if they crash

gaearon opened the pull request:
Previously we used to push them only after the instance was available. This caused issues in cases an error is thrown during `componentWillMount()`.

In that case we never got to pushing the provider in the begin phase, but in complete phase the provider check returned `true` since the instance existed by that point. As a result we got mismatching context pops.

We solve the issue by making the context check independent of whether the instance actually exists. Instead we're checking the type itself.

This lets us push class context early. However there's another problem: we might not know the context value. If the instance is not yet created, we can't call `getChildContext` on it.

To fix this, we are introducing a way to replace current value on the stack, and a way to read the previous value. This also helps remove some branching and split the memoized from invalidated code paths.

I also added a test from #8604 to verify this also solves the problem described in that PR.

gaearon committed:
Push class context providers early

Previously we used to push them only after the instance was available. This caused issues in cases an error is thrown during componentWillMount().

In that case we never got to pushing the provider in the begin phase, but in complete phase the provider check returned true since the instance existed by that point. As a result we got mismatching context pops.

We solve the issue by making the context check independent of whether the instance actually exists. Instead we're checking the type itself.

This lets us push class context early. However there's another problem: we might not know the context value. If the instance is not yet created, we can't call getChildContext on it.

To fix this, we are introducing a way to replace current value on the stack, and a way to read the previous value. This also helps remove some branching and split the memoized from invalidated code paths.

gaearon committed:
Add a now-passing test from #8604

Also rename another test to have a shorter name.

gaearon committed:
Move isContextProvider() checks into push() and pop()

All uses of push() and pop() are guarded by it anyway.

This makes it more similar to how we use host context.

There is only one other place where isContextProvider() is used and that's legacy code needed for renderSubtree().

gaearon commented:
There is one additional bug there but I can't figure out how to fix it.

When an error boundary itself is a context provider, `beginFailedWork` doesn't push it.
If I make it push it though, I get unexpected pops in unwinding.
I don't understand how this part of the code works but maybe @acdlite has ideas.

bvaughn commented on src/renderers/shared/fiber/ReactFiberContext.js:
```diff
@@ -38,22 +40,24 @@ if (__DEV__) {
-  const unmaskedContext = getUnmaskedContext();
-  const context = {};
+  const hasOwnContext = isContextProvider(workInProgress);
+  // If it is a context provider then use the previous context instead.
```
nit: This comment makes sense to me, with the context I have, but might not be sufficient explanation for someone seeing this part of the code for the first time. Maybe we should elaborate about _why_ we use the previous one. Maybe even just a pointer to see the comments in `pushContextProvider`?

bvaughn commented on src/renderers/shared/fiber/ReactFiberContext.js:
```diff
@@ -71,14 +75,15 @@ exports.hasContextChanged = function() : boolean {
-    // Instance might be null, if the fiber errored during construction
-    fiber.stateNode &&
-    typeof fiber.stateNode.getChildContext === 'function'
+    fiber.type.childContextTypes != null
```
👍

bvaughn commented on src/renderers/shared/fiber/ReactFiberContext.js:
```diff
@@ -117,21 +122,34 @@ function processChildContext(fiber : Fiber, parentContext : Object, isReconcilin
+exports.invalidateContextProvider = function(workInProgress : Fiber) : void {
+  const instance = workInProgress.stateNode;
+  if (instance == null) {
+    throw new Error('Expected to have an instance by this point.');
```
Curious why you're using `throw` here instead of `invariant`? (How do we determine when to do one vs the other?)

gaearon replied:
We should always use invariants but I didn't bother because we're going to do a separate pass to replace them all anyway. (It's in the umbrella.)

bvaughn commented on src/renderers/shared/fiber/ReactFiberStack.js:
```diff
@@ -34,6 +34,21 @@ exports.createCursor = function<T>(defaultValue : T) : StackCursor<T> {
+  if (index === 0) {
+    return null;
+  }
+  return valueStack[index - 1];
```
Isn't previous actually the value at `valueStack[index]`?

Current is in `cursor.current` and previous is the top item in the stack (at `index`).

bvaughn replied:
PS I was already thinking about doing this, but I believe I'll submit a follow-up PR that adds some unit tests for `ReactFiberStack` itself.

gaearon replied:
Ah, I think you're right. I wonder why tests didn't catch this.

bvaughn replied:
No problem. I'll add some tests for this specific type of thing. 👍

bvaughn commented on src/renderers/shared/fiber/ReactFiberStack.js:
```diff
@@ -34,6 +34,21 @@ exports.createCursor = function<T>(defaultValue : T) : StackCursor<T> {
+    }
+    return null;
+  }
+  if (index === 0) {
```
This check is unnecessary. We only care about < 0 which is already checked for above.

bvaughn replied:
Rather than "unnecessary" I should have said "bad" because it would prevent the _default_ value from being returned as previous if one was set. (eg like the default for `didPerformWorkStackCursor`)

gaearon committed:
Clarify why we read the previous context

gaearon committed:
Use invariant instead of throwing an error

gaearon committed:
Fix off-by-one in ReactFiberStack

gaearon commented:
Addressed.

bvaughn commented:
Nice clean up Dan!

gaearon commented:
My only concern if c57cf3ef6f18960026b18c7d5981259d89060f7e is bad from optimization point of view. I.e. maybe it prevents inlining or something. What do you think?

gaearon commented:
Hmm maybe off by one wasn't a bug after all..

bvaughn commented:
> My only concern if c57cf3e is bad from optimization point of view. I.e. maybe it prevents inlining or something. What do you think?

Unfortunately this is not something I'm that knowledgable about yet. I should defer to @sebmarkbage.

> Hmm maybe off by one wasn't a bug after all..

You mean `getPrevious`? It definitely was.

gaearon commented:
Then there must be another off-by-one there that cancels it out since the test started failing after I fixed it (it goes into an infinite loop and shows wrong context in some tests).

bvaughn commented:
Hm. I'll finish up these `ReactFiberStack` tests and then I can help diagnose.

bvaughn commented:
Tests added for `ReactFiberStack` ~> https://github.com/gaearon/react/pull/1

gaearon commented:
Hmm, I just realized `valueStack` and `fiberStack` aren't "tracking" each other despite both being indexed by the same index. `fiberStack[i]` corresponds to the current value whereas `valueStack[i]` is one value behind.

gaearon commented:
This was a little confusing to me. While we're at it: why doesn't `reset` reset the current value (but resets its fiber)?

bvaughn commented:
> why doesn't reset `reset` the current value (but resets its fiber)

The stack doesn't track its allocated cursors so it has no way to reset current values. (`ReactFiberContext` and `ReactFiberHostContext` are responsible for doing this.) In the future- if we change the context API- we might be allocating a lot of cursors so it seemed best not to track them/hold onto them. (I was kind of on the fence about this to be honest.)

Once the stack has been reset, any cursor values should be ignored as they are no longer safe. If someone chooses to hold onto one, then they can though.

gaearon commented:
I totally misunderstood how stack works.
I thought values were held separately. 😄

bvaughn commented:
FYI in `ReactIncremental` test...
* This test fails: _merges and masks context_
* This test hangs: _does not leak own context into context provider_

gaearon commented:
My approach worked by accident. By addressing `index - 1` I just managed to get the right value (e.g. context rather than `didWork`) from the stack but it just happened to work.

bvaughn commented:
Ha! That's funny 😁  Yeah I had a few moments of off-by-one confusion like that when working on the initial stack branch.

bvaughn commented:
I totally should have spotted that. Sorry.

This type of stack-management stuff is still a bit foreign to my thinking.

gaearon commented:
Do I understand correctly that order of `push` and `pop` operations must always be exact reverse or we risk popping the wrong thing corresponding to the right fiber?

bvaughn commented:
Yes. That is correct. Mirror order.

gaearon commented:
It might make sense to track cursor objects together with fibers in DEV for better checks.

bvaughn reviewed (changes requested):
We need an alternate approach to `getPrevious`. Sorry for not spotting this initially. ☹️

gaearon commented:
I don't think replace can work on anything but innermost level, right? We can't replace two things one after another. We can only pop twice and push twice.

bvaughn commented:
You can `replace` multiple cursors, so long as you're only concerned with updating the "current" value of each. (Or you could pop+push again yes.) `replace` checks for a Fiber match when but that's not technically needed. You _could_ push another cursor/Fiber, and still safely `replace` a value from a previous cursor/Fiber's push, so long as it's the top-most value for that cursor.

The part about this that doesn't work is getting the previous value. Stack currently doesn't know or care about what's on it or in what order, so the previous value is not guaranteed to be for the same cursor. We'd have to track previous on the cursor itself if we wanted to be able to return it. Alternately we could mirror pop-then-repush but this assumes that nothing else has pushed to the stack in the meanwhile.

bvaughn commented:
> We'd have to track previous on the cursor itself if we wanted to be able to return it.

Actually even this would be problematic, because once we pop- we'd have the same problem, trying to figure what the previous was.

gaearon committed:
Manually keep track of the last parent context

The previous algorithm was flawed and worked by accident, as shown by the failing tests after an off-by-one was fixed.

The implementation of getPrevious() was incorrect because the context stack currently has no notion of a previous value per cursor.

Instead, we are caching the previous value directly in the ReactFiberContext in a local variable.

Additionally, we are using push() and pop() instead of adding a new replace() method.

gaearon commented:
OK I think that fixes it.

bvaughn reviewed (approved):
This looks good!

gaearon commented:
I'll land since it fixes bugs, happy to follow up on perf improvements if @sebmarkbage sees any.
